### PR TITLE
Rearranged Lambert solvers and resolving conflicting files

### PR DESCRIPTION
**Pull request by MaqOwais**
Tried to make a better version according to latest updates as well as removing some errors accordingly

#14972

NO ENTRY

### CODE REVIEW

**Comment by MaqOwais** [comment]
> I'll be soon discussing the summary about this changes I've made, shortly !

File: sympy/solvers/solveset.py [hunk 1]
```diff
@@ -987,7 +991,7 @@ def _solveset(f, symbol, domain, _check=False):
                         result_rational = _solve_as_rational(equation, symbol, domain)
                         if isinstance(result_rational, ConditionSet):
                             # may be a transcendental type equation
-                            result += _transolve(equation, symbol, domain)
+                            result = _transolve(equation, symbol, domain)
```
Comment (MaqOwais) [inline sympy/solvers/solveset.py]: Case(1)

File: sympy/solvers/solveset.py [hunk 2]
```diff
@@ -1026,6 +1030,11 @@ def _solveset(f, symbol, domain, _check=False):
             result = FiniteSet(*[s for s in result
                       if isinstance(s, RootOf)
                       or domain_check(fx, symbol, s)])
+        elif type(result) is list:
+            for _ in range(len(result)-1):
+                result = result[_] + result[_+1]
+            if len(result) == 1:
+                return result[0]
```
Comment (MaqOwais) [inline sympy/solvers/solveset.py]: Case(1)
These are the respective lines of code for error modification.

File: sympy/solvers/bivariate.py [hunk 3]
```diff
@@ -179,8 +179,25 @@ def _lambert(eq, x):
...
+            j += 1
+            de = "{}".format(i)
+            if 'I' not in de:
+                ind_ls.append(j)
+        for i in ind_ls:
+            args.append(args1[i])
+    else:
+        args = [d/(a*b)*t for t in roots(t**p - e, t).keys() ]
+    if len(args) == 0:
+        return S.EmptySet
```
Comment (MaqOwais) [inline sympy/solvers/bivariate.py]: Case(2)
Also the argument part as domain is included at _lambert , _solve_lambert , _solve_even_degree_expr to solve in real domain

File: sympy/solvers/bivariate.py [hunk 4]
```diff
@@ -228,7 +249,7 @@ def _solve_lambert(f, symbol, gens):
       X = B, a = -1, d = a*log(p), f = -log(d) - g*log(p)
     """
 
-    def _solve_even_degree_expr(expr, t, symbol):
+    def _solve_even_degree_expr(expr, t, symbol,domain= S.Complexes):
```
Comment (MaqOwais) [inline sympy/solvers/bivariate.py]: case(2)

**Comment by MaqOwais** [comment]
From #14792 here's the finishing touch up. 
Here with cases we are going to discuss more below accordingly.

### Case(1):
**Here we see an error message specifically in `.subs` case so, here I modified accordingly and finally arrived to required solution.**
### Input :
```
solveset_real((x**2 - 2*x + 1).subs(x, log(x) + 3*x), x)
solveset_real((x**2 - 2*x + 1).subs(x, (log(x) + 3*x)**2 - 1), x)
solveset_real((x**2 - 2*x - 2).subs(x, log(x) + 3*x), x) 
```
### Output:
> Traceback (most recent call last):
>   File "<stdin>", line 1, in <module>
>   File "/Users/owais/Desktop/symp_doc/sympy/sympy/solvers/solveset.py", line 2127, in solveset_real
>     return solveset(f, symbol, S.Reals)
>   File "/Users/owais/Desktop/symp_doc/sympy/sympy/solvers/solveset.py", line 2105, in solveset
>     return solveset(f.xreplace({symbol: r}), r, domain
>   File "/Users/owais/Desktop/symp_doc/sympy/sympy/solvers/solveset.py", line 2123, in solveset
>     return _solveset(f, symbol, domain, _check=True)
>   File "/Users/owais/Desktop/symp_doc/sympy/sympy/solvers/solveset.py", line 994, in _solveset
>     result += _transolve(equation, symbol, domain)
> TypeError: unsupported operand type(s) for +=: 'EmptySet' and 'list'`

### After Modification :
```
FiniteSet(LambertW(3*E)/3)

FiniteSet(LambertW(3*exp(sqrt(2)))/3, LambertW(3*exp(-sqrt(2)))/3)

FiniteSet(LambertW(3*exp(1 + sqrt(3)))/3, LambertW(3*exp(1 - sqrt(3)))/3)
```

### Case(2):
**Here, It was returning the complex solutions, but the requirement is only to get real solution hence accordingly modified.**
### PART(A):
### Input:
```
solveset_real((a/x + exp(x/2)).diff(x, 2), x)

solveset_real(3*x + 5 + 2**(-5*x + 3), x)     
```
### Output:

> Complement(FiniteSet(6*LambertW((-a)**(1/3)/3), 6*LambertW(-(-a)**(1/3)/6 - sqrt(3)*I*(-a)**(1/3)/6), 6*LambertW(-(-a)**(1/3)/6 + sqrt(3)*I*(-a)**(1/3)/6)), FiniteSet(0))
> FiniteSet(-5/3 + LambertW(-10240*2**(1/3)*log(2)/3)/(5*log(2)), -5/3 + 
> LambertW(-5*(-1024*2**(1/3) - 1024*2**(1/3)*sqrt(3)*I)*log(2)/3)/(5*log(2)), -5/3 + 
> LambertW(-5*(-1024*2**(1/3) + 1024*2**(1/3)*sqrt(3)*I)*log(2)/3)/(5*log(2)))

### After Modification:
```
Complement(FiniteSet(6*LambertW((-a)**(1/3)/3)), FiniteSet(0))
FiniteSet(-5/3 + LambertW(-10240*2**(1/3)*log(2)/3)/(5*log(2)))
```
### PART(B):
### Input:
```
`p = symbols('p', positive=True)`
`eq = 3*log(p**(3*x + 5)) + p**(3*x + 5)`
`solveset(eq, x)`
`solveset_real(eq, x)`
```
### Output:
> FiniteSet(-5/3 - LambertW(1/3)/(3*log(p)))
> FiniteSet(-5/3 - LambertW(1/3)/(3*log(p)), -5/3 - LambertW(-1/6 - sqrt(3)*I/6)/(3*log(p)), -5/3 - LambertW(-1/6 + sqrt(3)*I/6)/(3*log(p)))`
### After Modification: 
```
FiniteSet(-5/3 - LambertW(1/3)/(3*log(p)))
FiniteSet(-5/3 - LambertW(1/3)/(3*log(p)))
```
### PART(C):
### Input :
```
solveset_real(2*x + 5 + log(3*x - 2), x)
```
### Output:
> FiniteSet(LambertW(2*exp(-19/3)/3)/2 + 2/3, 2/3 + LambertW(-exp(-19/3)/3 - sqrt(3)*I*exp(-19/3)/3)/2, 2/3 + LambertW(-exp(-19/3)/3 + sqrt(3)*I*exp(-19/3)/3)/2)

### After Modification:
```
FiniteSet(LambertW(2*exp(-19/3)/3)/2 + 2/3)
```

**I'm really fascinated by the work on Lambert equations(obviously huge credit specifically goes to  @smichr 
 @Yathartha22 @jmig5776 and other's ) as it solves more such varieties of problems, hence increasing the domain of solving the equations in more specific manner.**

_But some more such equations are to be solved in `solveset_real,` where the solutions are obtained by `solve` method but instead we are not getting it with `solveset_real` so more work has to be done considering all cases ._

I'm really glad to have further discussions and also to work on it.

**Comment by Yathartha22** [comment]
Are there any changes that you took from #14972? If so, it is always better to cherry-pick the commits to retain the authorship.
@jmig5776 can you look over the changes, I will try to look by this weekend.

**Comment by MaqOwais** [comment]
> Are there any changes that you took from #14972? If so, it is always better to cherry-pick the commits to retain the authorship.

Oh !
Actually ,I'm very new to git so not aware of such functionalities(cherry-pick).

**Comment by oscarbenjamin** [comment]
@MaqOwais there are many commits here that are one line changes which suggests that you are trying to fix things by changing one line at a time and then pushing to see the tests run on Travis.

Do you know how to run the tests locally on your computer? It's much faster to run the tests locally and fix them before pushing because then you don't have to wait for Travis. It's also a much better use of Travis time which is a limited resource for the project.

I would do this sort of thing by e.g. running (pip install pytest pytest-xdist) first
```
$ pytest sympy/solvers -m 'not slow' -n4 -d
```
The `-n4` means use 4 cores so adjust for however many cores you have.

Then once you can see that some tests have failed you can do
```
$ pytest sympy/solvers -m 'not slow' -n4 -d --lf
```
where the `--lf` means run the last-failed tests. This means you can run only the tests that didn't pass first time round which makes it quick to fix them one at a time. Once you've got all the tests passed with `--lf` you can then run the lot again without `--lf`. If they all pass then that might be a good time to commit and push.

**Comment by MaqOwais** [comment]
Thank you so much for your guidance... 
Actually I'm a newbie ...

I'll definitely not repeat this again..

**Comment by MaqOwais** [comment]
@oscarbenjamin your strategy is really effective .
It helped me a lot ..
Thanks for your attention and sorry of the inconvenience and chaos I've done over here.  : )

File: sympy/solvers/bivariate.py [hunk 5]
```diff
@@ -179,7 +179,23 @@ def _lambert(eq, x):
...
+        t = e
+        args = [d/(a*b)*t]
+    elif domain.is_subset(S.Reals):
+        ind_ls = [d/(a*b)*t for t in roots(t**p - e, t).keys() ]
+        args = []
+        j = -1
+        for i in ind_ls:
+            j += 1
+            de = "{}".format(i)
+            if 'I' not in de:
```
Comment (smichr) [inline sympy/solvers/bivariate.py]: -> `if not i.has(I):`

File: sympy/solvers/bivariate.py [hunk 6]
```diff
@@ -268,13 +288,14 @@ def _solve_even_degree_expr(expr, t, symbol):
         ``-x``. So the role of the ``t`` in the expression received by
         this function is to mark where ``+/-x`` should be inserted
         before obtaining the Lambert solutions.
-
         """
         nlhs, plhs = [
             expr.xreplace({t: sgn*symbol}) for sgn in (-1, 1)]
-        sols = _solve_lambert(nlhs, symbol, gens)
+        sols = _solve_lambert(nlhs, symbol, gens,domain)
```
Comment (smichr) [inline sympy/solvers/bivariate.py]: space after commas (as for those that came before)

File: sympy/solvers/solveset.py [hunk 7]
```diff
@@ -1008,7 +1012,44 @@ def _solveset(f, symbol, domain, _check=False):
...
+                            if len(indls) == 1:
+                                if indls[0] == 0:
+                                    return result
+                        if indls is not None:
+                            args = []
+                            j = -1
+                            for _ in indls:
+                                j += 1
+                                de = "{}".format(_)
+                                if 'I' not in de:
```
Comment (smichr) [inline sympy/solvers/solveset.py]: -> `if not i.has(I):`

File: sympy/solvers/solveset.py [hunk 8]
```diff
@@ -1652,6 +1698,77 @@ def _is_logarithmic(f, symbol):
...
+    if gpu is not None:
+        g, p, u = gpu
+        if g != lhs:
+            usol = _solveset(p, u, domain)
+            if not isinstance(usol, ConditionSet):
+                result = [_solveset(g - us, symbol, domain) for us in usol]
+    return result
+
+
+def _is_lambert(f, symbol):
```
Comment (smichr) [inline sympy/solvers/solveset.py]: see comment on original PR regarding this name and docstring

File: sympy/solvers/solveset.py [hunk 9]
```diff
@@ -1008,7 +1012,44 @@ def _solveset(f, symbol, domain, _check=False):
...
+                        'Abs' in str(f):
+                        return result
+                    elif not domain.is_subset(S.Integers) and \
+                        ((result.has(cos) and not result.has(sin)) or \
+                            (result.has(sin) and not result.has(cos))) :
+                        f = solve(f,x)
+                        f = FiniteSet(*f)
+                        return f
+                elif domain.is_subset(S.Reals):
+                    if not 'sqrt' in str(f) and len(str(f)) < 120:
```
Comment (smichr) [inline sympy/solvers/solveset.py]: Please don't use strings to do the test. You can do `if not any(i for i in f.atoms(Pow) if i.exp is S.Half) and f.count_ops() < smthng:`

Though I am not sure why there is a limit on the length/complexity of `f`.

**Comment by smichr** [comment]
> sorry of the inconvenience and chaos I've done over here

an easy way to get this PR back in line with the orginal is to do create a local copy of the other PR, say `lamb`. Then
```
git checkout lamb
git merge lamb master
git checkout patch-3
git merge patch-3 master
git diff patch-3 lamb > dif
git branch backup
git checkout lamb
git branch -D patch-3
git checkout -b patch-3
git apply dif
```
Then commit the changes. The patch-3 will have all the new additions to the old PRs branch. Then push here with the `-f` option:

```
git push -f MaqOwais patch-3
```

File: sympy/solvers/solveset.py [hunk 10]
```diff
@@ -1026,6 +1068,11 @@ def _solveset(f, symbol, domain, _check=False):
             result = FiniteSet(*[s for s in result
                       if isinstance(s, RootOf)
                       or domain_check(fx, symbol, s)])
+        elif isinstance(result,list):
+            for _ in range(len(result)-1):
```
Comment (czgdp1807) [inline sympy/solvers/solveset.py]:
```suggestion
            for _ in range(len(result) - 1):
```

**Comment by czgdp1807** [comment]
@MaqOwais Some comments aren't addressed here. Please look into them.

File: sympy/solvers/solveset.py (hunk 10 above)
Comment (MaqOwais) [reply sympy/solvers/solveset.py:1072]: > Some comments aren't addressed here. Please look into them.

I think this is only the last change :)

**Comment by MaqOwais** [comment]
@czgdp1807 
what's the reason for not accepting the commit ?
Everything is, I think, is all correct..

**Comment by czgdp1807** [comment]
@smichr Can we merge this?

**Pull request closed without merging**